src/http/conf.rs

```diff
@@ -1,63 +1,281 @@
+        let conf_ctx = self.ctx.cast::<ngx_http_conf_ctx_t>();
+        let conf_ctx = conf_ctx.as_ref()?;
+        NonNull::new((*conf_ctx.srv_conf.add(module.ctx_index)).cast())
+    }
+
+    #[inline]
+    unsafe fn http_location_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf_ctx = self.ctx.cast::<ngx_http_conf_ctx_t>();
+        let conf_ctx = conf_ctx.as_ref()?;
+        NonNull::new((*conf_ctx.loc_conf.add(module.ctx_index)).cast())
+    }
 }
 
-/// # Safety
+impl HttpModuleConfExt for ngx_http_core_srv_conf_t {
+    #[inline]
+    unsafe fn http_main_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf_ctx = self.ctx.as_ref()?;
+        NonNull::new((*conf_ctx.main_conf.add(module.ctx_index)).cast())
+    }
+
+    #[inline]
+    unsafe fn http_server_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf_ctx = self.ctx.as_ref()?;
+        NonNull::new((*conf_ctx.srv_conf.add(module.ctx_index)).cast())
+    }
+
+    #[inline]
+    unsafe fn http_location_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf_ctx = self.ctx.as_ref()?;
+        NonNull::new((*conf_ctx.loc_conf.add(module.ctx_index)).cast())
+    }
+}
+
+impl HttpModuleConfExt for ngx_http_request_t {
+    #[inline]
+    unsafe fn http_main_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        NonNull::new((*self.main_conf.add(module.ctx_index)).cast())
+    }
+
+    #[inline]
+    unsafe fn http_server_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        NonNull::new((*self.srv_conf.add(module.ctx_index)).cast())
+    }
+
+    #[inline]
+    unsafe fn http_location_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        NonNull::new((*self.loc_conf.add(module.ctx_index)).cast())
+    }
+}
+
+impl HttpModuleConfExt for ngx_http_upstream_srv_conf_t {
+    #[inline]
+    unsafe fn http_server_conf_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf = self.srv_conf;
+        if conf.is_null() {
+            return None;
+        }
+        NonNull::new((*conf.add(module.ctx_index)).cast())
+    }
+}
+
+/// Trait to define and access main module configuration
 ///
-/// The caller has provided a value `ngx_http_upstream_srv_conf_t. If the `us` argument is null, a
-/// None Option is returned; however, if the `us` internal fields are invalid or the module index
-/// is out of bounds failures may still occur.
-pub unsafe fn ngx_http_conf_upstream_srv_conf_immutable<T>(
-    us: *const ngx_http_upstream_srv_conf_t,
-    module: &ngx_module_t,
-) -> Option<*const T> {
-    if us.is_null() {
-        return None;
+/// # Safety
+/// Caller of configuration access methods must ensure that pointers to configuration structures
+/// inside objects used as parameters are NULL or refer to valid configuration data.
```

Comment:
We are lifting the safety requirements of `HttpModuleConfExt::http_main_conf_unchecked` to the trait level, so it should inherit the safety invariant.
Everything else is already guaranteed by the `ngx_http_block` code or our implementations of `HttpModuleConfExt`.

```suggestion
/// Caller must ensure that type `HttpModuleMainConf::MainConf` matches the configuration type
/// for the specified module.
```